File: server/src/objective-result.ts
export const OBJECTIVE_RESULT_MARKER = 'NSPI_RESULT_V1:';

export type ObjectiveQuestionKind =
  | 'single_choice' | 'multiple_choice' | 'ordering' | 'short_fill' | 'other';
export type ObjectiveResultState = 'ready' | 'review' | 'retake';
export type ObjectiveResultReason =
  | 'none' | 'ambiguous_question' | 'ambiguous_options' | 'cropped' | 'unreadable'
  | 'missing_context' | 'unsupported';
export type ObjectiveParserPath = 'v1' | 'legacy_fallback' | 'legacy' | 'none';
export type ObjectiveProtocolViolation =
  | 'duplicateMarker' | 'markerNotLast' | 'invalidJSON' | 'oversizedJSON' | 'unknownField'
  | 'invalidEnum' | 'invalidStateCombination' | 'finalMismatch' | 'missingUsableResult';

export interface ObjectiveResultV1 {
  v: 1;
  kind: ObjectiveQuestionKind;
  state: ObjectiveResultState;
  answer: string | null;
  reason: ObjectiveResultReason;
}

export interface ObjectiveResultComposition {
  visibleText: string;
  finalAnswer: string | null;
  result: ObjectiveResultV1 | null;
  state: ObjectiveResultState | null;
  parserPath: ObjectiveParserPath;
  violations: ObjectiveProtocolViolation[];
}

const KINDS = new Set<ObjectiveQuestionKind>([
  'single_choice', 'multiple_choice', 'ordering', 'short_fill', 'other',
]);
const STATES = new Set<ObjectiveResultState>(['ready', 'review', 'retake']);
const REASONS = new Set<ObjectiveResultReason>([
  'none', 'ambiguous_question', 'ambiguous_options', 'cropped', 'unreadable',
  'missing_context', 'unsupported',
]);
const KEYS = ['answer', 'kind', 'reason', 'state', 'v'];
const OBJECTIVE_KINDS = new Set<ObjectiveQuestionKind>([
  'single_choice', 'multiple_choice', 'ordering', 'short_fill',
]);

function pushUnique(out: ObjectiveProtocolViolation[], value: ObjectiveProtocolViolation): void {
  if (!out.includes(value)) out.push(value);
}

function markerIsInsideMarkdownFence(lines: string[], markerIndex: number): boolean {
  let fence: '```' | '~~~' | null = null;
  for (const line of lines.slice(0, markerIndex)) {
    const match = /^\s*(```|~~~)/u.exec(line);
    if (!match) continue;
    const token = match[1] as '```' | '~~~';
    if (fence === null) fence = token;
    else if (fence === token) fence = null;
  }
  return fence !== null;
}

/** Shared FINAL/answer canonicalization. JS normalize('NFKC') is Unicode NFKC. */
export function normalizeObjectiveAnswer(value: string): string {
  let out = value.normalize('NFKC').trim().replace(/\s+/gu, ' ');
  let changed = true;
  while (changed) {
    changed = false;
    for (const pair of [['**', '**'], ['__', '__'], ['`', '`']] as const) {
      if (out.startsWith(pair[0]) && out.endsWith(pair[1]) && out.length > pair[0].length + pair[1].length) {
        out = out.slice(pair[0].length, -pair[1].length).trim();
        changed = true;
      }
    }
  }
  return /^[a-z]$/i.test(out) ? out.toUpperCase() : out;
}

export function extractFinalAnswer(text: string): string | null {
  const lines = text.replace(/\r\n?/g, '\n').split('\n');
  for (let i = lines.length - 1; i >= 0; i -= 1) {
    const line = lines[i]!;
    const match = /^[ \t]{0,3}(?:#{1,6}[ \t]*)?(?:\*{1,2}|_{1,2})?[ \t]*FINAL[ \t]*[:：][ \t]*(.+)$/iu.exec(line);
    if (!match) continue;
    let answer = match[1]!.trim();
    for (const fence of ['**', '__']) {
      const count = answer.split(fence).length - 1;
      if (count % 2 === 1) {
        if (answer.startsWith(fence)) answer = answer.slice(fence.length).trim();
        else if (answer.endsWith(fence)) answer = answer.slice(0, -fence.length).trim();
      }
    }
    return answer.length > 0 && [...answer].length <= 512 ? answer : null;
  }
  return null;
}

function validateObject(raw: unknown, violations: ObjectiveProtocolViolation[]): ObjectiveResultV1 | null {
  if (typeof raw !== 'object' || raw === null || Array.isArray(raw)) {
    pushUnique(violations, 'invalidJSON');
    return null;
  }
  const obj = raw as Record<string, unknown>;
  const keys = Object.keys(obj).sort();
  if (keys.length !== KEYS.length || keys.some((key, i) => key !== KEYS[i])) {
    pushUnique(violations, 'unknownField');
    return null;
  }
  if (obj.v !== 1 || typeof obj.kind !== 'string' || !KINDS.has(obj.kind as ObjectiveQuestionKind)
      || typeof obj.state !== 'string' || !STATES.has(obj.state as ObjectiveResultState)
      || typeof obj.reason !== 'string' || !REASONS.has(obj.reason as ObjectiveResultReason)) {
    pushUnique(violations, 'invalidEnum');
    return null;
  }
  const result = obj as unknown as ObjectiveResultV1;
  const answerLength = typeof result.answer === 'string' ? [...result.answer].length : 0;
  let valid = true;
  if (result.state === 'ready') {
    valid = typeof result.answer === 'string' && answerLength >= 1 && answerLength <= 512
      && result.reason === 'none' && OBJECTIVE_KINDS.has(result.kind);
  } else if (result.state === 'review') {
    valid = typeof result.answer === 'string' && answerLength >= 1 && answerLength <= 512
      && ['ambiguous_question', 'ambiguous_options', 'missing_context', 'unsupported'].includes(result.reason)
      && (result.kind !== 'other' || result.reason === 'unsupported');
  } else {
    valid = result.answer === null && ['cropped', 'unreadable', 'missing_context'].includes(result.reason);
  }
  if (!valid) {
    pushUnique(violations, 'invalidStateCombination');
    return null;
  }
  return result;
}

/** Strict completion-time parser. `protocolEnabled=false` preserves legacy FINAL behavior. */
export function composeObjectiveResult(text: string, protocolEnabled = true): ObjectiveResultComposition {
  const normalizedText = text.replace(/\r\n?/g, '\n');
  const lines = normalizedText.split('\n');
  const markerIndexes = lines.flatMap((line, index) => line.startsWith(OBJECTIVE_RESULT_MARKER) ? [index] : []);
  const violations: ObjectiveProtocolViolation[] = [];
  if (markerIndexes.length > 1) pushUnique(violations, 'duplicateMarker');
  const nonEmptyLast = lines.findLastIndex((line) => line.trim() !== '');
  if (markerIndexes.length > 0 && markerIndexes.at(-1) !== nonEmptyLast) pushUnique(violations, 'markerNotLast');

  const visibleText = lines
    .filter((line) => !line.startsWith(OBJECTIVE_RESULT_MARKER))
    .join('\n')
    .replace(/\s+$/u, '');
  const finalAnswer = extractFinalAnswer(visibleText);
  let result: ObjectiveResultV1 | null = null;

  const markerInFence = markerIndexes.length === 1
    && markerIsInsideMarkdownFence(lines, markerIndexes[0]!);
  if (markerInFence) pushUnique(violations, 'invalidStateCombination');

  if (protocolEnabled && markerIndexes.length === 1 && markerIndexes[0] === nonEmptyLast && !markerInFence) {
    const payload = lines[markerIndexes[0]!]!.slice(OBJECTIVE_RESULT_MARKER.length).trimStart();
    if (Buffer.byteLength(payload, 'utf8') > 4096) {
      pushUnique(violations, 'oversizedJSON');
    } else {
      try {
        result = validateObject(JSON.parse(payload) as unknown, violations);
      } catch {
        pushUnique(violations, 'invalidJSON');
      }
    }
    if (result && result.state !== 'retake') {
      if (!finalAnswer || normalizeObjectiveAnswer(finalAnswer) !== normalizeObjectiveAnswer(result.answer ?? '')) {
        pushUnique(violations, 'finalMismatch');
        result = null;
      }
    } else if (result?.state === 'retake' && finalAnswer !== null) {
      pushUnique(violations, 'invalidStateCombination');
      // An explicitly unusable result must not become a chargeable FINAL fallback.
      // The contradictory envelope is still invalid; do not turn it into a valid V1 score.
      pushUnique(violations, 'missingUsableResult');
      return { visibleText: '', finalAnswer: null, result: null, state: null, parserPath: 'none', violations };
    }
  }

  if (result) {
    return { visibleText, finalAnswer: result.answer, result, state: result.state, parserPath: 'v1', violations };
  }
  if (finalAnswer) {
    return {
      visibleText, finalAnswer, result: null,
      state: protocolEnabled ? 'review' : null,
      parserPath: protocolEnabled ? 'legacy_fallback' : 'legacy', violations,
    };
  }
  pushUnique(violations, 'missingUsableResult');
  return { visibleText, finalAnswer: null, result: null, state: null, parserPath: 'none', violations };
}

export function objectiveResultIsBillable(composition: ObjectiveResultComposition): boolean {
  return composition.parserPath === 'legacy_fallback'
    || (composition.parserPath === 'v1' && composition.state !== 'retake');
}

File: Sources/NotchSPI/Notch/ObjectiveResult.swift
import Foundation

enum ObjectiveQuestionKind: String, Codable, CaseIterable {
    case singleChoice = "single_choice"
    case multipleChoice = "multiple_choice"
    case ordering
    case shortFill = "short_fill"
    case other
}

enum ObjectiveResultState: String, Codable {
    case ready, review, retake
}

enum ObjectiveResultReason: String, Codable {
    case none
    case ambiguousQuestion = "ambiguous_question"
    case ambiguousOptions = "ambiguous_options"
    case cropped, unreadable
    case missingContext = "missing_context"
    case unsupported
}

struct ObjectiveResultV1: Codable, Equatable {
    let v: Int
    let kind: ObjectiveQuestionKind
    let state: ObjectiveResultState
    let answer: String?
    let reason: ObjectiveResultReason
}

enum ObjectiveParserPath: String, Codable {
    case v1
    case legacyFallback = "legacy_fallback"
    case legacy
    case none
}

enum ObjectiveProtocolViolation: String, Codable, Equatable {
    case duplicateMarker, markerNotLast, invalidJSON, oversizedJSON, unknownField
    case invalidEnum, invalidStateCombination, finalMismatch, missingUsableResult
}

struct ObjectiveResultComposition: Equatable {
    let visibleText: String
    let finalAnswer: String?
    let result: ObjectiveResultV1?
    let state: ObjectiveResultState?
    let parserPath: ObjectiveParserPath
    let violations: [ObjectiveProtocolViolation]
    var noResultReason: String? = nil
}

enum ObjectiveResultParser {
    static let marker = "NSPI_RESULT_V1:"
    private static let keys = Set(["v", "kind", "state", "answer", "reason"])
    private static let objectiveKinds = Set(ObjectiveQuestionKind.allCases.filter { $0 != .other })

    static func normalizeAnswer(_ source: String) -> String {
        var value = source.precomposedStringWithCompatibilityMapping
            .trimmingCharacters(in: .whitespacesAndNewlines)
        value = value.replacingOccurrences(of: #"\s+"#, with: " ", options: .regularExpression)
        var changed = true
        while changed {
            changed = false
            for fence in ["**", "__", "`"] where value.hasPrefix(fence) && value.hasSuffix(fence)
                && value.count > fence.count * 2 {
                value = String(value.dropFirst(fence.count).dropLast(fence.count))
                    .trimmingCharacters(in: .whitespacesAndNewlines)
                changed = true
            }
        }
        if value.range(of: #"^[A-Za-z]$"#, options: .regularExpression) != nil {
            return value.uppercased()
        }
        return value
    }

    static func compose(raw: String, protocolEnabled: Bool = true) -> ObjectiveResultComposition {
        if raw.contains("NSPI_NO_RESULT_V1:") {
            let reason = ScreenQueryDiagnostic.parse(raw)
            return .init(visibleText: "", finalAnswer: nil, result: nil, state: nil, parserPath: .none,
                         violations: reason == nil ? [.invalidJSON] : [], noResultReason: reason)
        }
        let normalized = raw.replacingOccurrences(of: "\r\n", with: "\n")
            .replacingOccurrences(of: "\r", with: "\n")
        let lines = normalized.components(separatedBy: "\n")
        let markerIndexes = lines.indices.filter { lines[$0].hasPrefix(marker) }
        let lastNonempty = lines.indices.last { !lines[$0].trimmingCharacters(in: .whitespaces).isEmpty }
        var violations: [ObjectiveProtocolViolation] = []
        func append(_ violation: ObjectiveProtocolViolation) {
            if !violations.contains(violation) { violations.append(violation) }
        }
        if markerIndexes.count > 1 { append(.duplicateMarker) }
        if let last = markerIndexes.last, last != lastNonempty { append(.markerNotLast) }
        let markerInFence = markerIndexes.count == 1 && isInsideMarkdownFence(
            lines: lines, markerIndex: markerIndexes[0])
        if markerInFence { append(.invalidStateCombination) }

        let visible = lines.filter { !$0.hasPrefix(marker) }.joined(separator: "\n")
            .trimmingCharacters(in: .whitespacesAndNewlines)
        let final = AnswerComposer.parse(visible, streaming: false).final.flatMap {
            let value = $0.trimmingCharacters(in: .whitespacesAndNewlines)
            return !value.isEmpty && value.unicodeScalars.count <= 512 ? value : nil
        }
        var result: ObjectiveResultV1?

        if protocolEnabled, markerIndexes.count == 1, markerIndexes[0] == lastNonempty,
           !markerInFence {
            let line = lines[markerIndexes[0]]
            let payload = String(line.dropFirst(marker.count)).drop(while: { $0 == " " || $0 == "\t" })
            if payload.utf8.count > 4_096 {
                append(.oversizedJSON)
            } else if let data = String(payload).data(using: .utf8),
                      let object = try? JSONSerialization.jsonObject(with: data),
                      let dictionary = object as? [String: Any] {
                if Set(dictionary.keys) != keys {
                    append(.unknownField)
                } else if let v = dictionary["v"] as? Int, v == 1,
                          let kindRaw = dictionary["kind"] as? String,
                          let kind = ObjectiveQuestionKind(rawValue: kindRaw),
                          let stateRaw = dictionary["state"] as? String,
                          let state = ObjectiveResultState(rawValue: stateRaw),
                          let reasonRaw = dictionary["reason"] as? String,
                          let reason = ObjectiveResultReason(rawValue: reasonRaw),
                          dictionary["answer"] is String || dictionary["answer"] is NSNull {
                    let answer = dictionary["answer"] as? String
                    let candidate = ObjectiveResultV1(v: v, kind: kind, state: state, answer: answer, reason: reason)
                    if valid(candidate) { result = candidate } else { append(.invalidStateCombination) }
                } else {
                    append(.invalidEnum)
                }
            } else {
                append(.invalidJSON)
            }

            if let candidate = result, candidate.state != .retake {
                if final == nil || normalizeAnswer(final!) != normalizeAnswer(candidate.answer ?? "") {
                    append(.finalMismatch)
                    result = nil
                }
            } else if result?.state == .retake, final != nil {
                append(.invalidStateCombination)
                // Preserve the unusable result instead of offering FINAL as an answer.
                append(.missingUsableResult)
                return .init(visibleText: "", finalAnswer: nil, result: nil, state: nil,
                             parserPath: .none, violations: violations)
            }
        }

        if let result {
            return .init(visibleText: visible, finalAnswer: result.answer, result: result,
                         state: result.state, parserPath: .v1, violations: violations)
        }
        if let final {
            let fallback = protocolEnabled
            return .init(visibleText: visible, finalAnswer: final, result: nil,
                         state: fallback ? .review : nil,
                         parserPath: fallback ? .legacyFallback : .legacy,
                         violations: violations)
        }
        append(.missingUsableResult)
        return .init(visibleText: visible, finalAnswer: nil, result: nil, state: nil,
                     parserPath: .none, violations: violations)
    }

    private static func valid(_ value: ObjectiveResultV1) -> Bool {
        let answerValid = value.answer.map { !$0.isEmpty && $0.unicodeScalars.count <= 512 } ?? false
        switch value.state {
        case .ready:
            return answerValid && value.reason == .none && objectiveKinds.contains(value.kind)
        case .review:
            let validReasons: Set<ObjectiveResultReason> = [
                .ambiguousQuestion, .ambiguousOptions, .missingContext, .unsupported,
            ]
            return answerValid && validReasons.contains(value.reason)
                && (value.kind != .other || value.reason == .unsupported)
        case .retake:
            return value.answer == nil && [.cropped, .unreadable, .missingContext].contains(value.reason)
        }
    }

    private static func isInsideMarkdownFence(lines: [String], markerIndex: Int) -> Bool {
        var openFence: String?
        for line in lines.prefix(markerIndex) {
            let trimmed = line.trimmingCharacters(in: .whitespaces)
            let token = trimmed.hasPrefix("```") ? "```" : (trimmed.hasPrefix("~~~") ? "~~~" : nil)
            guard let token else { continue }
            if openFence == nil { openFence = token }
            else if openFence == token { openFence = nil }
        }
        return openFence != nil
    }
}

/// Incremental presentation filter. It never exposes a marker line, including partial prefixes.
struct ObjectiveResultStreamFilter {
    private(set) var rawBuffer = ""
    private var overflow = false

    mutating func append(_ delta: String) -> String {
        guard !overflow, rawBuffer.utf8.count + delta.utf8.count <= 64 * 1024 else {
            overflow = true; rawBuffer = ""; return ""
        }
        rawBuffer += delta
        return visible(streaming: true)
    }

    mutating func finish() -> ObjectiveResultComposition {
        ObjectiveResultParser.compose(raw: rawBuffer)
    }

    private func visible(streaming: Bool) -> String {
        let normalized = rawBuffer.replacingOccurrences(of: "\r\n", with: "\n")
            .replacingOccurrences(of: "\r", with: "\n")
        var lines = normalized.components(separatedBy: "\n")
        let openLast = !normalized.hasSuffix("\n")
        lines = lines.enumerated().compactMap { index, line in
            let markers = [ObjectiveResultParser.marker, "NSPI_NO_RESULT_V1:"]
            if markers.contains(where: { line.contains($0) }) { return nil }
            if streaming && openLast && index == lines.count - 1 {
                let candidate = line.drop(while: { $0 == " " || $0 == "\t" })
                if !candidate.isEmpty && markers.contains(where: { $0.hasPrefix(String(candidate)) }) { return nil }
            }
            return line
        }
        return lines.joined(separator: "\n")
    }
}


enum ScreenQueryDiagnostic {
    private struct Payload: Decodable { let v: Int; let reason: String }
    static func parse(_ text: String) -> String? {
        let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
        let marker = "NSPI_NO_RESULT_V1:"
        guard trimmed.hasPrefix(marker) else { return nil }
        let payload = String(trimmed.dropFirst(marker.count)).trimmingCharacters(in: .whitespaces)
        guard !payload.contains("\n"), !payload.contains("\r"), payload.utf8.count <= 4096,
              let data = payload.data(using: .utf8),
              let object = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              Set(object.keys) == Set(["v", "reason"]),
              let regex = try? NSRegularExpression(pattern: #""([^"\\]*)"\s*:"#),
              regex.numberOfMatches(in: payload, range: NSRange(payload.startIndex..., in: payload)) == 2,
              let decoded = try? JSONDecoder().decode(Payload.self, from: data), decoded.v == 1,
              ["unsupported_scope", "multiple_targets"].contains(decoded.reason) else { return nil }
        return decoded.reason
    }
}
